src/libsync/propagatecommonzsync.cpp

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright (C) by Ahmed Ammar <ahmed.a.ammar@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "config.h"
+#include "propagateupload.h"
+#include "owncloudpropagator_p.h"
+#include "networkjobs.h"
+#include "account.h"
+#include "common/syncjournaldb.h"
+#include "common/syncjournalfilerecord.h"
+#include "common/utility.h"
+#include "filesystem.h"
+#include "propagatorjobs.h"
+#include "syncengine.h"
+#include "propagateremotemove.h"
+#include "propagateremotedelete.h"
+#include "common/asserts.h"
+
+#include <QNetworkAccessManager>
+#include <QFileInfo>
+#include <QDir>
+#include <cmath>
+#include <cstring>
+#include <QTemporaryDir>
+
+#ifdef Q_OS_UNIX
+#include <unistd.h>
+#include <arpa/inet.h>
+#endif
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcZsyncSeed, "sync.propagate.zsync.seed", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncGenerate, "sync.propagate.zsync.generate", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncGet, "sync.networkjob.zsync.get", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncPut, "sync.networkjob.zsync.put", QtInfoMsg)
+
+bool isZsyncPropagationEnabled(OwncloudPropagator *propagator, SyncFileItemPtr item)
+{
+    if (!propagator->account()->capabilities().zsyncSupported()) {
+        qCInfo(lcPropagator) << "[zsync disabled] Lack of server support.";
+        return false;
+    }
+    if (item->_remotePerm.hasPermission(RemotePermissions::IsMounted) ||
+        item->_remotePerm.hasPermission(RemotePermissions::IsMountedSub)) {
+        qCInfo(lcPropagator) << "[zsync disabled] External storage not supported.";
```

Comment:
why not?  if the server says it has zsync enabled why should it not work?

Reply:
Because external drives are out of oC control, if they change then the metadata will be completely out-of-sync and corruption will ensue.

Reply:
Then oC should not set that zsync is enabled for these.

Reply:
Hmm, it's not that simple, oC doesn't have that much control over the process, the client has more, oC will not put the metadata if the client sends it for a file stored externally, but by that time the client has already done the cpu-intensive tasks of generating the metadata file and sent it. The client saves a lot by not doing that in the first place ...